web3/main.py

```diff
@@ -337,21 +343,23 @@ def attach_modules(
         """
         _attach_modules(self, modules)
 
-    def isConnected(self) -> bool:
+    def isConnected(self) -> Union[bool, Coroutine[Any, Any, bool]]:
         return self.provider.isConnected()
 
     def is_encodable(self, _type: TypeStr, value: Any) -> bool:
         return self.codec.is_encodable(_type, value)
 
     @property
-    def ens(self) -> ENS:
+    def ens(self) -> Union[ENS, AsyncENS]:
         if self._ens is cast(ENS, empty):
             return ENS.fromWeb3(self)
-        else:
-            return self._ens
+        elif self._ens is cast(AsyncENS, empty) or self.eth.is_async:
+            return AsyncENS.fromWeb3(self)
+
+        return self._ens
 
     @ens.setter
-    def ens(self, new_ens: ENS) -> None:
+    def ens(self, new_ens: Optional[Union[ENS, AsyncENS]]) -> None:
```

Comment:
It's a bit tricky to track all the ways this works but this [line leads](https://github.com/ethereum/web3.py/blob/master/ens/utils.py#L78) me to believe that it can be basically `None`, though only internally. Though I also think we can just not explicitly set it to `None` there and it will just be set to `cast(ENS, empty)`.

Any opinion on what you think would be the ideal setup here? 🤔